custom/ilsgateway/zipline/data_sources/zipline_warehouse_package_data_source.py

```diff
@@ -0,0 +1,58 @@
+from custom.ilsgateway.zipline.data_sources.zipline_data_source import ZiplineDataSource
+from custom.zipline.models import EmergencyOrderPackage
+
+from custom.ilsgateway.zipline import helpers
+
+class ZiplineWarehousePackageDataSource(ZiplineDataSource):
+
+    @property
+    def order_id(self):
+        return self.config.order_id
+
+    def get_emergency_order_packages(self, start, limit):
+        offset = start + limit
+        additional_filters = {}
+        if self.sql_location.location_type_object.administrative:
+            descendants = self.sql_location.get_descendants() \
+                .exclude(is_archived=True).values_list('site_code', flat=True)
+            additional_filters['order__location_code__in'] = descendants
+        else:
+            additional_filters['order__location_code'] = self.sql_location.site_code
+
+        if self.statuses:
+            additional_filters['status__in'] = self.statuses
+
+        if self.order_id:
+            additional_filters['order_id'] = self.order_id
+
+        return EmergencyOrderPackage.objects.filter(
+            order__domain=self.domain,
+            order__timestamp__range=[self.start_date, self.end_date],
+            **additional_filters
+        )[start:offset]
```

Comment:
also this can use select_related for the order, dispatched_status, and delivered_status foreign keys